client/src/components/campaignCreatorComponents/CampaignCreator.tsx

```diff
@@ -36,8 +43,81 @@ const CampaignCreator = ({ campaignSettings, setCampaignSettings }: CampaignCrea
         navigate('/twilight-library/dashboard');
     };
 
+    const addToTimeline = (
+        timeline: Record<number, string[]>,
+        nodeKey: keyof NodePillarLists,
+        typeKey: keyof TimelineOptionList,
+    ) => {
+        if (!selectedCampaign.flexible_nemesis_encounters && typeKey === 'nemesis') {
+            // Handle strict Nemesis timeline
+            const nemesis_tier: string = nodeKey.slice(-1);
+            selectedCampaign[nodeKey].forEach((selection: string) => {
+                [1, 2, 3].forEach((level: number) => {
+                    const fight_str: string = `nn${nemesis_tier}_lvl${level}_fight_year`;
+                    const fight_key: keyof NemesisFightYearLists = fight_str as keyof NemesisFightYearLists;
+                    const fight_year: number | null = selectedCampaign[fight_key];
+                    if (fight_year !== null) {
+                        timeline[fight_year].push(`NE - ${selection} lvl ${level}`);
+                    }
+                });
+            });
+            return;
+        }
+
+        selectedCampaign[nodeKey].forEach((selection: string) => {
+            const query: Record<string, Record<string, string[]>> = campaignOptionsData.timeline[typeKey];
+            if (query.hasOwnProperty(selection)) {
+                Object.keys(query[selection]).forEach((yearKey: string) => {
+                    query[selection][Number(yearKey)].forEach((yearData: string) => {
+                        timeline[Number(yearKey)].push(yearData);
+                    });
+                });
+            }
+        });
+    };
+
     const handleCreateTimeline = (event: React.FormEvent<HTMLFormElement>) => {
```

Comment:
Okay, so this function is incredibly hard to understand wtf is happening.  The reason is, you're thinking of manipulating this data completely upside down.

You're going with "here's a set of instructions on what to do to this".  Instead, it's often more useful to thing about building something like this one item at a time.  
So instead you would say "what is in year one, let's build that first, then year 2, etc."
Or, you can "decorate", which is when you have an object, and a bunch of operations that change that object in one clear way.  This is what you're doing with `add to timeline` but you need to make ALL of this decorations, the main function should just be "Here's an object, and here's all the decorators I am adding, by calling wlel named functions that tell em WHAT they are doing. Then if I care HOW, I can click into them"

Reply:
refactored it. i think i am following your suggested change.